substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/LinkAtBuildTimeFeature.java

```diff
@@ -0,0 +1,169 @@
+         * the synthetic modifier set (clazz.isSynthetic() returns false for such classes). Any
+         * class with package-name jdk.internal.reflect should be treated as link-at-build-time.
+         */
+        requireCompletePackageOrClass.put("jdk.internal.reflect", null);
+
+        Options.LinkAtBuildTime.getValue().getValuesWithOrigins().forEach(this::extractOptionValue);
+
+        ImageSingletons.add(LinkAtBuildTimeSupport.class, new LinkAtBuildTimeSupport(this));
+    }
+
+    private void extractOptionValue(Pair<String, OptionOrigin> valueOrigin) {
+        var value = valueOrigin.getLeft();
+        OptionOrigin origin = valueOrigin.getRight();
+        if (value.isEmpty()) {
+            if (origin.commandLineLike()) {
+                requireCompleteAll = true;
+                return;
+            }
+            var originModule = uriModuleMap.get(origin.container());
+            if (originModule != null) {
+                requireCompleteModules.add(originModule);
+                return;
+            }
+            throw UserError.abort("Using '%s' without args only allowed on module-path. %s not part of module-path.",
+                            SubstrateOptionsParser.commandArgument(Options.LinkAtBuildTime, value), origin);
+        } else {
+            for (String entry : OptionUtils.resolveOptionValueRedirection(Options.LinkAtBuildTime, value, origin)) {
+                if (validOptionValue.matcher(entry).matches()) {
+                    requireCompletePackageOrClass.computeIfAbsent(entry, unused -> new HashSet<>()).add(origin);
+                } else {
+                    throw UserError.abort("Entry '%s' in option '%s' provided by '%s' is neither a package nor a fully qualified classname.",
+                                    entry, SubstrateOptionsParser.commandArgument(Options.LinkAtBuildTime, value), origin);
+                }
+            }
+        }
+    }
+
+    boolean linkAtBuildTime(Class<?> clazz) {
+        return linkAtBuildTimeImpl(clazz) != null;
+    }
+
+    @SuppressWarnings("unchecked")
+    String linkAtBuildTimeReason(Class<?> clazz) {
+        Object reason = linkAtBuildTimeImpl(clazz);
+        if (reason == null) {
+            return null;
+        }
+        if (reason instanceof String) {
+            return (String) reason;
+        }
+        Set<OptionOrigin> origins = (Set<OptionOrigin>) reason;
+        return origins.stream().map(OptionOrigin::toString).collect(Collectors.joining(" and "));
+    }
+
+    private Object linkAtBuildTimeImpl(Class<?> clazz) {
+        if (requireCompleteAll) {
+            return reasonCommandLine;
+        }
+
+        if (clazz.isArray() || !classLoaderSupport.isNativeImageClassLoader(clazz.getClassLoader())) {
+            return "system default";
+        }
+        assert !clazz.isPrimitive() : "Primitive classes are not loaded via NativeImageClassLoader";
```

Comment:
I would put this at the top of the method.

Reply:
I can't. The `!classLoaderSupport.isNativeImageClassLoader(...)`-bit is the thing that ensures primitives are seen as `"system default"`.